test/e2e/cluster_upgrade.go

```diff
@@ -0,0 +1,122 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Skipped", func() {
+	Describe("Cluster upgrade", func() {
+		name := "baz"
+		framework := Framework{BaseName: "cluster-upgrade"}
+		var webserver *WebserverTest
+
+		BeforeEach(func() {
+			framework.beforeEach()
+			webserver = NewWebserverTest(framework.Client, framework.Namespace.Name, name)
+			webserver.CreateWebserverPod()
+			svc := webserver.BuildServiceSpec()
+			svc.Spec.Type = api.ServiceTypeLoadBalancer
+			webserver.CreateService(svc)
+		})
+
+		AfterEach(func() {
+			framework.afterEach()
+			webserver.Cleanup()
+		})
+
+		It("by kube-push of master should maintain responsive services", func() {
+			testClusterUpgrade(framework, name, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-push.sh", "-m")
+			})
+		})
+
+		It("by gce-upgrade of master should maintain responsive services", func() {
+			if !providerIs("gce") {
+				By(fmt.Sprintf("Skippingt test, which is not implemented for %s", testContext.Provider))
+				return
+			}
+			testClusterUpgrade(framework, name, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-upgrade.sh", "-M", "-l")
+			})
+		})
+	})
+})
+
+func testClusterUpgrade(framework Framework, svcName string, upgrade func()) {
+	result, err := waitForLoadBalancerIngress(framework.Client, svcName, framework.Namespace.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if len(result.Status.LoadBalancer.Ingress) != 1 {
+		Failf("got unexpected number (%v) of ingress points for externally load balanced service: %v", result.Status.LoadBalancer.Ingress, result)
+	}
+	ingress := result.Status.LoadBalancer.Ingress[0]
+	By("Waiting for pod to become reachable")
+	testLoadBalancerReachable(ingress, 80)
+
+	By("Upgrading cluster")
+	httpClient := http.Client{Timeout: 2 * time.Second}
+	retries := 3
+	done := make(chan struct{}, 1)
+	go util.Until(func() {
+		defer GinkgoRecover()
+		var err error
+
+		ip := ingress.IP
+		if ip == "" {
+			ip = ingress.Hostname
+		}
+
+		for i := 0; i < retries; i++ {
+			r, err := httpClient.Get("http://" + ip)
+			if err != nil {
+				continue
+			}
+			if r.StatusCode < http.StatusOK || r.StatusCode >= http.StatusNotFound {
+				err = fmt.Errorf("received bad status code: %v")
+			}
+			return
+		}
+		Failf("health check failed after %v retries: %v", retries, err)
+	}, 200*time.Millisecond, done)
+
+	By("starting upgrade")
+	upgrade()
+
+	done <- struct{}{}
```

Comment:
Could you please add a bit more thorough validation of the existence of things? The service polling should generally take care of these assertions at a high level (i.e. if things are working, the backing objects probably exist), but I'd really like a bit more low-level checking that the following exist as we expect after an upgrade:
- replication controllers
- pods
- services

Some kind of `validate()` function would be awesome for this purpose.

I guess this is getting at the broader notion that we don't just want to test that a service works during an upgrade, but that the objects we created before an upgrade are all running and healthy afterwards.